### PR TITLE
ENH: Add startswith & endswith ufuncs for unicode and bytes dtypes

### PR DESCRIPTION
**Pull request by lysnikolaou**

### CODE REVIEW

**Comment by ngoldbaum** [comment]
I haven’t looked closely at the other PRs yet but I appreciate that you’re using a character-oriented low-level API in this. It should be straightforward to adapt this to a utf-8 string. 

If you can, it would be nice if you can avoid creating a low-level API in this string ufunc that accepts an array of unicode code points, as that will require an unnecessary allocation if I want to use it with stringdtype.

**Comment by ngoldbaum** [comment]
Thinking about me adapting these ufuncs to work with stringdtype, I guess I need to convert the strings to UTF-32 before using this API?

Is it possible to structure things so that it will be straightforward to support UTF-8 as well? It’s straightforward to generate a stream of UTF-32 codepoints from a UTF-8 string and in principle there’s no need to allocate a buffer to store the UTF-32 representation. Maybe some kind of approach based on a stream of characters?

Another idea, instead of the low-level API taking a pointer to an array of code points, make it always take bytes, and then have some kind of templated function that consumes bytes and returns code points?

Just some thoughts! I see you working on this and it makes me a little anxious because every string ufunc that gets added is more work for me eventually and it would be nice if we could do that in a way that minimizes future pain.

**Comment by seberg** [comment]
One general comment:  For all of these new PRs, it would be good to consolidate the boilerplate at least for the functions that have exactly the same pattern.

I suspect it depends a bit on what you are doing, for these functions the `string_endswith` and `string_startswith` helpers should work just fine?  But when things like whitespace gets involved, you may have to embrace UCS4 buffers.

**Comment by ngoldbaum** [comment]
I'm not a big fan of C++ streams, so how about making the implementation use a template function that returns a UCS4 character? For unicode and bytes it'll be trivial but for UTF-8 it'll be [this function](https://github.com/numpy/numpy-user-dtypes/blob/b5b7ca2d43c176ee517b7c0042c93018f5fe00d9/stringdtype/stringdtype/src/casts.c#L313) (looking at it now I should probably factor out the length check into the caller).

> Do we have the low-level API return a stream of UCS4 characters itself and the caller needs to add these to the output buffer after conversion?

For stringdtype specifically, I could straightforwardly add a low-level function that accepts a UTF-8 string and a stream of UCS4 characters and returns a new UTF-8 string that appends the characters from the UCS4 string to the original string without an additional intermediate allocation, so if your API returns an array of UCS4 characters and an integer size that would work.

**Comment by lysnikolaou** [comment]
@ngoldbaum Can you take a look at the last commit in #24868, where I tried to implement something like what we discussed above?

**Comment by lysnikolaou** [comment]
@seberg @ngoldbaum This is next, whenever you get the time. It should be a fairly straightforward review.

@ngoldbaum The review comments you left in the other PR that got merged have been addressed here.

**Review by ngoldbaum — COMMENTED** [review]
Overall looks great, just one substantive comment about documenting where the logic in `adjust_indices` comes from.

File: numpy/_core/src/umath/string_ufuncs.cpp [hunk 1]
```diff
@@ -55,6 +60,63 @@ string_rstrip(const character *str, int elsize)
 }
 
 
+/* helper to fixup start/end slice values */
```
[the remark this answers is not in the document]
Comment (lysnikolaou) [reply numpy/_core/src/umath/string_ufuncs.cpp:63]: Oops, sorry about that!

**Comment by ngoldbaum** [comment]
Thanks @lysnikolaou!

**Pull request merged**